app/scripts/services/builds.js

```diff
@@ -145,6 +146,57 @@ angular.module("openshiftConsole")
       return moment(endTimestamp).diff(moment(startTimestamp));
     };
 
+    var incompleteBuilds = function(builds) {
+      return _.map(builds, function(build) {
+        return isIncompleteBuild(build);
+      });
+    };
+
+    var completeBuilds = function(builds) {
+      return _.map(builds, function(build) {
+        return !isIncompleteBuild(build);
+      });
+    };
+
+    var lastCompleteByBuildConfig = function(builds) {
+      return _.reduce(
+                completeBuilds(builds),
```

Comment:
there are far more complete builds than incomplete builds, implementing it this way means you are iterating over almost the entire list of builds in the namespace twice, just in this one function, and this happens anytime there is an update to any build.